internal/website/content/howto/pubsub/publish.md

```diff
@@ -0,0 +1,410 @@
+[SNS]: https://aws.amazon.com/sns/
+
+### Amazon Simple Notification Service Constructor {#sns-ctor}
+
+The [`awssnssqs.OpenTopic`][] constructor opens an SNS topic. You must first
+create an [AWS session][] with the same region as your topic:
+
+```go
+import (
+    "github.com/aws/aws-sdk-go/aws"
+    "github.com/aws/aws-sdk-go/aws/session"
+    "github.com/aws/aws-sdk-go/service/sns"
+    "gocloud.dev/pubsub/awssnssqs"
+)
+// ...
+
+// Establish an AWS session.
+// The region must match the region for "MyTopic".
+sess, err := session.NewSession(&aws.Config{
+    Region: aws.String("us-east-2"),
+})
+if err != nil {
+    return err
+}
+
+// Create a *pubsub.Topic.
+client := sns.New(sess)
+const topicARN = "arn:aws:sns:us-east-2:123456789012:MyTopic"
+topic, err := awssnssqs.OpenTopic(ctx, client, topicARN, nil)
+if err != nil {
+    return err
+}
+defer topic.Shutdown(ctx)
+```
+
+[`awssnssqs.OpenTopic`]: https://godoc.org/gocloud.dev/pubsub/awssnssqs#OpenTopic
+[AWS session]: https://docs.aws.amazon.com/sdk-for-go/api/aws/session/
+
+## Google Cloud Pub/Sub {#gcp}
+
+The Go CDK can publish to a Google [Cloud Pub/Sub][] topic. The URLs use the
+project ID and the topic ID. `pubsub.OpenTopic` will use [Application Default
+Credentials][GCP creds].
+
+```go
+import (
+    "gocloud.dev/pubsub"
+    _ "gocloud.dev/pubsub/gcppubsub"
+)
+
+// ...
+
+topic, err := pubsub.OpenTopic(ctx,
+    "gcppubsub://my-project/my-topic")
+if err != nil {
+    return err
+}
+defer topic.Shutdown(ctx)
+```
+
+[Cloud Pub/Sub]: https://cloud.google.com/pubsub/docs/
+[GCP creds]: https://cloud.google.com/docs/authentication/production
+
+### Google Cloud Pub/Sub Constructor {#gcp-ctor}
+
+The [`gcppubsub.OpenTopic`][] constructor opens a Cloud Pub/Sub topic. You
+must first obtain [GCP credentials][GCP creds] and then create a gRPC
+connection to Cloud Pub/Sub. (This gRPC connection can be reused among
+topics.)
+
+```go
+import (
+    "gocloud.dev/gcp"
+    "gocloud.dev/pubsub/gcppubsub"
+)
+
+// ...
+
+// Your GCP credentials.
+// See https://cloud.google.com/docs/authentication/production
+// for more info on alternatives.
+creds, err := gcp.DefaultCredentials(ctx)
+if err != nil {
+    return err
+}
+// Get the project ID from the credentials (required by OpenTopic).
+projectID, err := gcp.DefaultProjectID(creds)
+if err != nil {
+    return err
+}
+
+// Open a gRPC connection to the GCP Pub/Sub API.
+conn, cleanup, err := gcppubsub.Dial(ctx, creds.TokenSource)
+if err != nil {
+    return err
+}
+defer cleanup()
+
+// Construct a PublisherClient using the connection.
+pubClient, err := gcppubsub.PublisherClient(ctx, conn)
+if err != nil {
+    return err
+}
+defer pubClient.Close()
+
+// Construct a *pubsub.Topic.
+topic := gcppubsub.OpenTopic(pubClient, projectID, "example-topic", nil)
+defer topic.Shutdown(ctx)
+```
+
+[`gcppubsub.OpenTopic`]: https://godoc.org/gocloud.dev/pubsub/gcppubsub#OpenTopic
+
+## Azure Service Bus {#azure}
```

Comment:
I don't want to be annoying, but why are we duplicating all of this provider-specific documentation here? Shouldn't it be in the godoc/examples?

Reply:
Because godoc is reference, not a how-to guide. Quoting [What nobody tells you about documentation](https://www.divio.com/blog/documentation/):

> Reference guides are _technical descriptions of the machinery_ and how to operate it.
>
> Reference guides have one job only: to describe. They are code-determined, because ultimately that's what they describe: key classes, functions, APIs, and so they should list things like functions, fields, attributes and methods, and set out how to use them.
>
> Reference material is **information-oriented.**
>
> By all means technical reference can contain examples to illustrate usage, but it should not attempt to explain basic concepts, or how to achieve achieve common tasks.

This is the format I've been doing for the other sections. Rationale:

1.  It's not straight duplication: there is a fair amount of editing that I'm doing as I go to clean up artifacts of godoc examples like predeclaring variables and blank identifier assignments.
2.  Linking would be unsatisfactory: it would force the reader to jump to a number of different packages to understand how to do one task end-to-end.
3.  Developing a way to inline godoc examples would suffer from the aforementioned problem of needing to edit.

I think that the examples in godoc should likely become more focused to individual functions over time, rather than whole-package like they largely are now. Definitely we should link between these things more, but since the how-to guides aren't published, I don't want to drop broken links into production reference.

Reply:
Discussed offline. Resolution:

1. Example code should be embedded in how-to guides, but source-of-truth should be godoc. Keeping as-is for sake of submitting, but I'm going to solve #1827 before moving forward with other guides.
2. Reference docs for particular functions should stay where they are.
3. Larger concept discussion will move to the website, and be linked from the godoc. Tracking in #1469.